### PR TITLE
Stop relying on local API discovery cache.

### PR DESCRIPTION
**Pull request by madhusudancs**
`kubectl` caches API discovery result in a local directory and it's possible for the cache to go out of sync. There is no mechanism to invalidate the out-of-date local cache. It so happens that when clusters come up for the first time, `kubectl` gets an incomplete set of API group versions in API discovery results. `kubectl` still caches it but doesn't refresh it for 10 minutes. This affects `kubefed` which relies on `kubectl`'s libraries to perform API discovery since these libraries read from the local, incomplete cache.

In order to workaround this problem, we now use a discovery client that doesn't rely on the local cache.

Issue: #47977

**Release note**:
```release-note
NONE
```

/assign @csbell @perotinus 

/cc @pwittrock

### CODE REVIEW

**Comment by dchen1107** [comment]
Could you please xref that issue here?

**Comment by madhusudancs** [comment]
@dchen1107 done.

File: federation/pkg/kubefed/util/util.go [hunk 1]
```diff
@@ -259,15 +261,22 @@ func buildConfigFromSecret(secret *api.Secret, serverAddress string) (*restclien
 // clientset for either the preferred version or the first listed version (if no preference listed)
 // TODO: We need to evaluate the usage of RESTMapper interface to achieve te same functionality
 func GetVersionedClientForRBACOrFail(hostFactory cmdutil.Factory) (client.Interface, error) {
-	discoveryclient, err := hostFactory.DiscoveryClient()
```
Comment (liggitt) [inline federation/pkg/kubefed/util/util.go]: discoveryclient has `Fresh()` and `Invalidate()` functions that make it easy to tell if cached data was used, and request new data. I'd expect a usage like this:

```go
func GetVersionedClientForRBACOrFail(hostFactory cmdutil.Factory) (client.Interface, error) {
	discoveryclient, err := hostFactory.DiscoveryClient()
	if err != nil {
		return nil, err
	}
	rbacVersion, err := getRBACVersion(discoveryclient)
	if err != nil && !discoveryclient.Fresh() {
		discoveryclient.Invalidate()
		rbacVersion, err = getRBACVersion(discoveryclient)
	}
	if err != nil {
		return nil, err
	}
	return hostFactory.ClientSetForVersion(&gv)
}
func getRBACVersion(discoveryclient discovery.CachedDiscoveryInterface) (*schema.GroupVersion, error) {
	groupList, err := discoveryclient.ServerGroups()
	if err != nil {
		return nil, fmt.Errorf("Couldn't get clientset to create RBAC roles in the host cluster: %v", err)
 	}
	for _, g := range groupList.Groups {
		if g.Name == rbac.GroupName {
			if g.PreferredVersion.GroupVersion != "" {
				gv, err := schema.ParseGroupVersion(g.PreferredVersion.GroupVersion)
 				if err != nil {
 					return nil, err
 				}
 				return &gv, nil
 			}
			for _, version := range g.Versions {
				if version.GroupVersion != "" {
					gv, err := schema.ParseGroupVersion(version.GroupVersion)
					if err != nil {
						return nil, err
					}
					return &gv, nil
				}
			}
		}
	}
	return nil, &NoRBACAPIError{rbacAPINotAvailable}
}
```

**Comment by mbohlool** [comment]
if we implement etags for API discovery, we should be able to have a reliable cache: work in progress #47979, right?
also look at #48026. If we decide to go with a fix like that, we may not need this workaround.

**Comment by liggitt** [comment]
> also look at #48026. If we decide to go with a fix like that, we may not need this workaround. (also this workaround will not work if somebody uses kubectl 1.6).

this isn't a workaround, it's a bugfix for incorrect usage of the cached discovery client

> if we implement etags for API discovery, we should be able to have a reliable cache

sure, additional functionality could make the cached discovery client smarter.

File: federation/pkg/kubefed/util/util.go (hunk 1 above)
Comment (liggitt) [inline federation/pkg/kubefed/util/util.go]: opened https://github.com/kubernetes/kubernetes/pull/48077 with this impl as an alternative

**Comment by madhusudancs** [comment]
/close 

(superseded by #48077)

**Comment by madhusudancs** [comment]
> if we implement etags for API discovery, we should be able to have a reliable cache: work in progress #47979, right?
> also look at #48026. If we decide to go with a fix like that, we may not need this workaround. (also this workaround will not work if somebody uses kubectl 1.6).

Yeah, this might fix the current issue. Also, it doesn't really hurt that much if kubefed always uses uncached discovery information.

**Pull request closed without merging**